### ISSUE TITLE
Prometheus not cleaning up WAL files

### ISSUE THREAD

**Issue by walterdolce** [prometheus/prometheus#6408]
## Bug Report
**What did you do?**
Ran Prometheus for a while. At some point it got `Evicted`, then was unable to cleanup WAL files ever again.

**What did you expect to see?**
Prometheus to clean up the WAL files and move onto another checkpoint

**What did you see instead? Under which circumstances?**
Prometheus ran but stacked `000***` WAL files over time and got stuck on a fixed `checkpoint.00NNNN`. 

**Environment**
GKE

* System information:
Standard GKE nodes pool set

* Prometheus version:
```
prometheus, version 2.13.1 (branch: HEAD, revision: 6f92ce56053866194ae5937012c1bec40f1dd1d9)
  build user:       root@88e419aa1676
  build date:       20191017-13:15:01
  go version:       go1.13.1
```

* Alertmanager version:
N/A

* Prometheus configuration file:
```
---
rule_files:
  - /etc/prometheus/rules/*.yml
global:
  scrape_interval: 30s
  scrape_timeout: 10s
  evaluation_interval: 60s
alerting:
  alertmanagers:
  - static_configs:
    - targets:
      - 'alertmanager:9093'
scrape_configs:
  - job_name: prometheus
    scrape_interval: 5s
    static_configs:
      - targets:
          - 'localhost:9090'
  - job_name: cadvisor
    scrape_interval: 10s
    scrape_timeout: 10s
    scheme: https
    tls_config:
      insecure_skip_verify: true
    bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
    kubernetes_sd_configs:
      - role: node
    relabel_configs:
      - action: labelmap
        regex: __meta_kubernetes_node_label_cloud_google_com_gke_nodepool
        replacement: nodepool
      - action: labelmap
        regex: __meta_kubernetes_node_label_failure_domain_beta_kubernetes_io_(.+)
        replacement: $1
      - target_label: __address__
        replacement: kubernetes.default.svc:443
      - source_labels: [__meta_kubernetes_node_name]
        regex: (.+)
        target_label: __metrics_path__
        replacement: /api/v1/nodes/${1}/proxy/metrics/cadvisor
    metric_relabel_configs:
      - action: replace
        source_labels: [id]
        regex: '^/machine\.slice/machine-rkt\\x2d([^\\]+)\\.+/([^/]+)\.service$'
        target_label: rkt_container_name
        replacement: '${2}-${1}'
      - action: replace
        source_labels: [id]
        regex: '^/system\.slice/(.+)\.service$'
        target_label: systemd_service_name
        replacement: '${1}'
  - job_name: 'kubernetes-service-endpoints'
    kubernetes_sd_configs:
      - role: endpoints
    relabel_configs:
      - action: labelmap
        regex: __meta_kubernetes_pod_label_(app|team)
        replacement: $1
      - action: labelmap
        regex: __meta_kubernetes_pod_label_helm_sh_chart
        replacement: helm_chart
      - source_labels: [__meta_kubernetes_namespace]
        action: replace
        target_label: kubernetes_namespace
      - source_labels: [__meta_kubernetes_service_name]
        action: replace
        target_label: kubernetes_name
      - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_scrape]
        action: keep
        regex: true
      - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_scheme]
        action: replace
        target_label: __scheme__
        regex: (https?)
      - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_path]
        action: replace
        target_label: __metrics_path__
        regex: (.+)
      - source_labels: [__address__, __meta_kubernetes_service_annotation_prometheus_io_port]
        action: replace
        target_label: __address__
        regex: (.+)(?::\d+);(\d+)
        replacement: $1:$2
  - job_name: 'kubernetes-pods'
    kubernetes_sd_configs:
      - role: pod
    relabel_configs:
      - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_scrape]
        action: keep
        regex: true
      - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_path]
        action: replace
        target_label: __metrics_path__
        regex: (.+)
      - source_labels: [__address__, __meta_kubernetes_pod_annotation_prometheus_io_port]
        action: replace
        regex: ([^:]+)(?::\d+)?;(\d+)
        replacement: $1:$2
        target_label: __address__
      - action: labelmap
        regex: __meta_kubernetes_pod_label_(app|team)
        replacement: $1
      - action: labelmap
        regex: __meta_kubernetes_pod_label_helm_sh_chart
        replacement: helm_chart
      - source_labels: [__meta_kubernetes_namespace]
        action: replace
        target_label: kubernetes_namespace
      - source_labels: [__meta_kubernetes_pod_name]
        action: replace
        target_label: kubernetes_pod_name
  - job_name: 'kubernetes-nodes'
    kubernetes_sd_configs:
      - role: node
    tls_config:
      insecure_skip_verify: true
    bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
    relabel_configs:
      - action: labelmap
        regex: __meta_kubernetes_node_label_cloud_google_com_gke_nodepool
        replacement: nodepool
      - action: labelmap
        regex: __meta_kubernetes_node_label_failure_domain_beta_kubernetes_io_(.+)
        replacement: $1
      - target_label: __address__
        replacement: kubernetes.default.svc.cluster.local:443
      - target_label: __scheme__
        replacement: https
      - source_labels: [__meta_kubernetes_node_name]
        regex: (.+)
        target_label: __metrics_path__
        replacement: /api/v1/nodes/${1}/proxy/metrics
```

Prometheus Kubernetes deployment:
```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "9"
    kube-score/ignore: pod-networkpolicy,container-security-context
  labels:
    app: prometheus
    app.kubernetes.io/instance: prometheus
    app.kubernetes.io/managed-by: Tiller
    app.kubernetes.io/name: prometheus
    app.kubernetes.io/version: v2.13.1
    helm.sh/chart: prometheus-1.1.3
  name: prometheus
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: prometheus
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 0
    type: RollingUpdate
  template:
    metadata:
      annotations:
        checksum/config: d2a860ef522ce24f2fd335162678ed7a7401f7a7a45ac37e14ac245f7245f302
        checksum/rules: c625c283c4fd8c900eb9a30542c4fca388de76b0b1c792213d62988174b215ae  3b3b53018fe62da2a850a6f49f1867922c4f36c162deadf2897d16608a05e22c
      creationTimestamp: null
      labels:
        app: prometheus
        app.kubernetes.io/instance: prometheus
        app.kubernetes.io/managed-by: Tiller
        app.kubernetes.io/name: prometheus
        app.kubernetes.io/version: v2.13.1
        helm.sh/chart: prometheus-1.1.3
    spec:
      affinity:
        podAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchLabels:
                app: prometheus
            topologyKey: kubernetes.io/hostname
      containers:
      - args:
        - --alertmanager.notification-queue-capacity=10000
        - --alertmanager.timeout=10s
        - --config.file=/etc/prometheus/prometheus.yml
        - --log.format=logfmt
        - --log.level=info
        - --query.lookback-delta=5m
        - --query.max-concurrency=40
        - --query.max-samples=50000000
        - --query.timeout=2m
        - --rules.alert.for-grace-period=10m
        - --rules.alert.for-outage-tolerance=1h
        - --rules.alert.resend-delay=1m
        - --storage.remote.read-concurrent-limit=10
        - --storage.remote.read-sample-limit=5e7
        - --storage.tsdb.no-lockfile
        - --storage.tsdb.path=/prometheus/data/
        - --storage.tsdb.retention.time=30d
        - --storage.tsdb.wal-compression
        - --web.console.libraries=/etc/prometheus/console_libraries
        - --web.console.templates=/etc/prometheus/consoles
        - --web.max-connections=512
        - --web.read-timeout=1m
        - --web.route-prefix=/
        command:
        - prometheus
        image: prom/prometheus:v2.13.1
        imagePullPolicy: Always
        name: prometheus
        resources:
          requests:
            cpu: 250m
            memory: 300Mi
        securityContext:
          privileged: false
          readOnlyRootFilesystem: true
          runAsUser: 10001
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /tmp
          name: tmp-volume
        - mountPath: /etc/prometheus/rules
          name: prometheus-alerting-rules
          readOnly: true
        - mountPath: /etc/prometheus/prometheus.yml
          name: prometheus-config
          readOnly: true
          subPath: prometheus.yml
        - mountPath: /prometheus/data
          name: prometheus-data-disk
      dnsPolicy: ClusterFirst
      initContainers:
      - args:
        - -c
        - chown -R 10001:10002 /prometheus/data
        command:
        - sh
        image: prom/prometheus:v2.13.1
        imagePullPolicy: Always
        name: prometheus-data-folder-permissions-fixer
        resources:
          limits:
            cpu: 100m
            memory: 40Mi
          requests:
            cpu: 50m
            memory: 20Mi
        securityContext:
          privileged: false
          readOnlyRootFilesystem: true
          runAsUser: 0
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /prometheus/data
          name: prometheus-data-disk
      nodeSelector:    # Prometheus runs in a different node pool with more resources available to it
        prometheusx2: "true"
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: prometheus
      serviceAccountName: prometheus
      terminationGracePeriodSeconds: 30
      volumes:
      - emptyDir: {}
        name: tmp-volume
      - configMap:
          defaultMode: 420
          name: prometheus-alerting-rules-helm
        name: prometheus-alerting-rules    # Alerting rules, feel free to remove for this
      - configMap:
          defaultMode: 420
          name: prometheus-config-helm
        name: prometheus-config   # The prometheus config specified above ends up here
      - name: prometheus-data-disk
        persistentVolumeClaim:
          claimName: prometheus-data-disk    # This is a normal Google Cloud persistent disk, 100GB
```

* Alertmanager configuration file:
N/A

* Logs:
Prometheus logs: https://gist.github.com/walterdolce/2a8b1990234d0641c7e960a45383f08e
WAL folder content:
```
/prometheus $ ls -lah data/wal
total 10G
drwxr-xr-x    3 10001    10002       4.0K Dec  4 12:29 .
drwxr-xr-x   24 10001    10002       4.0K Dec  4 12:28 ..
-rw-r--r--    1 10001    10002     128.0M Nov 27 14:19 00002288
-rw-r--r--    1 10001    10002     127.9M Nov 27 14:58 00002289
-rw-r--r--    1 10001    10002       4.8M Nov 27 15:00 00002290
-rw-r--r--    1 10001    10002     128.0M Nov 27 15:39 00002291
-rw-r--r--    1 10001    10002     128.0M Nov 27 16:19 00002292
-rw-r--r--    1 10001    10002     128.0M Nov 27 16:58 00002293
-rw-r--r--    1 10001    10002       6.3M Nov 27 17:00 00002294
-rw-r--r--    1 10001    10002      61.1M Nov 27 17:19 00002295
-rw-r--r--    1 10001    10002      23.0M Nov 27 17:24 00002296
-rw-r--r--    1 10001    10002      94.6M Nov 27 17:53 00002297
-rw-r--r--    1 10001    10002     128.0M Nov 27 18:33 00002298
-rw-r--r--    1 10001    10002     127.9M Nov 27 19:13 00002299
-rw-r--r--    1 10001    10002       4.9M Nov 28 13:48 00002300
-rw-r--r--    1 10001    10002     128.0M Nov 28 14:27 00002301
-rw-r--r--    1 10001    10002      10.3M Nov 29 06:24 00002302
-rw-r--r--    1 10001    10002     127.6M Nov 29 07:05 00002303
-rw-r--r--    1 10001    10002     127.9M Nov 29 07:45 00002304
-rw-r--r--    1 10001    10002     128.0M Nov 29 08:26 00002305
-rw-r--r--    1 10001    10002     128.0M Nov 29 09:06 00002306
-rw-r--r--    1 10001    10002     127.9M Nov 29 09:46 00002307
-rw-r--r--    1 10001    10002     127.9M Nov 29 10:26 00002308
-rw-r--r--    1 10001    10002     127.8M Nov 29 11:06 00002309
-rw-r--r--    1 10001    10002     128.0M Nov 29 11:46 00002310
-rw-r--r--    1 10001    10002     127.9M Nov 29 12:26 00002311
-rw-r--r--    1 10001    10002     128.0M Nov 29 13:06 00002312
-rw-r--r--    1 10001    10002     127.7M Nov 29 13:47 00002313
-rw-r--r--    1 10001    10002     128.0M Nov 29 14:27 00002314
-rw-r--r--    1 10001    10002     128.0M Nov 29 15:07 00002315
-rw-r--r--    1 10001    10002     128.0M Nov 29 15:48 00002316
-rw-r--r--    1 10001    10002     127.6M Nov 29 16:28 00002317
-rw-r--r--    1 10001    10002     128.0M Nov 29 17:09 00002318
-rw-r--r--    1 10001    10002     128.0M Nov 29 17:49 00002319
-rw-r--r--    1 10001    10002     128.0M Nov 29 18:29 00002320
-rw-r--r--    1 10001    10002     128.0M Nov 29 19:10 00002321
-rw-r--r--    1 10001    10002     127.6M Nov 29 19:50 00002322
-rw-r--r--    1 10001    10002     128.0M Nov 29 20:31 00002323
-rw-r--r--    1 10001    10002     128.0M Nov 29 21:11 00002324
-rw-r--r--    1 10001    10002      63.6M Nov 29 21:31 00002325
-rw-r--r--    1 10001    10002      68.6M Nov 29 22:08 00002326
-rw-r--r--    1 10001    10002     127.9M Nov 29 22:59 00002327
-rw-r--r--    1 10001    10002     128.0M Nov 29 23:39 00002328
-rw-r--r--    1 10001    10002     128.0M Nov 30 00:20 00002329
-rw-r--r--    1 10001    10002     128.0M Nov 30 01:00 00002330
-rw-r--r--    1 10001    10002     128.0M Nov 30 01:41 00002331
-rw-r--r--    1 10001    10002     127.6M Nov 30 02:21 00002332
-rw-r--r--    1 10001    10002     128.0M Nov 30 03:02 00002333
-rw-r--r--    1 10001    10002     128.0M Nov 30 03:43 00002334
-rw-r--r--    1 10001    10002     127.9M Nov 30 04:24 00002335
-rw-r--r--    1 10001    10002     128.0M Nov 30 05:06 00002336
-rw-r--r--    1 10001    10002     127.7M Nov 30 05:47 00002337
-rw-r--r--    1 10001    10002     128.0M Nov 30 06:28 00002338
-rw-r--r--    1 10001    10002     128.0M Nov 30 07:09 00002339
-rw-r--r--    1 10001    10002     127.9M Dec  3 12:19 00002340
-rw-r--r--    1 10001    10002     128.0M Dec  3 12:58 00002341
-rw-r--r--    1 10001    10002     128.0M Dec  3 13:37 00002342
-rw-r--r--    1 10001    10002     128.0M Dec  3 14:16 00002343
-rw-r--r--    1 10001    10002     127.9M Dec  3 14:55 00002344
-rw-r--r--    1 10001    10002     128.0M Dec  3 15:33 00002345
-rw-r--r--    1 10001    10002     127.9M Dec  3 16:12 00002346
-rw-r--r--    1 10001    10002     128.0M Dec  3 16:50 00002347
-rw-r--r--    1 10001    10002     128.0M Dec  3 17:30 00002348
-rw-r--r--    1 10001    10002      43.2M Dec  3 17:43 00002349
-rw-r--r--    1 10001    10002     127.9M Dec  3 18:40 00002350
-rw-r--r--    1 10001    10002      28.2M Dec  3 18:48 00002351
-rw-r--r--    1 10001    10002          0 Dec  3 18:49 00002352
-rw-r--r--    1 10001    10002      47.5M Dec  3 19:51 00002353
-rw-r--r--    1 10001    10002          0 Dec  3 19:52 00002354
-rw-r--r--    1 10001    10002     128.0M Dec  3 21:17 00002355
-rw-r--r--    1 10001    10002     127.9M Dec  3 21:56 00002356
-rw-r--r--    1 10001    10002     127.9M Dec  3 22:35 00002357
-rw-r--r--    1 10001    10002     128.0M Dec  3 23:13 00002358
-rw-r--r--    1 10001    10002     127.8M Dec  3 23:52 00002359
-rw-r--r--    1 10001    10002     127.9M Dec  4 00:31 00002360
-rw-r--r--    1 10001    10002     128.0M Dec  4 01:09 00002361
-rw-r--r--    1 10001    10002     127.9M Dec  4 01:48 00002362
-rw-r--r--    1 10001    10002     128.0M Dec  4 02:27 00002363
-rw-r--r--    1 10001    10002     127.7M Dec  4 03:06 00002364
-rw-r--r--    1 10001    10002     128.0M Dec  4 03:44 00002365
-rw-r--r--    1 10001    10002     128.0M Dec  4 04:23 00002366
-rw-r--r--    1 10001    10002     127.9M Dec  4 05:02 00002367
-rw-r--r--    1 10001    10002     128.0M Dec  4 05:40 00002368
-rw-r--r--    1 10001    10002     127.8M Dec  4 06:19 00002369
-rw-r--r--    1 10001    10002     128.0M Dec  4 06:58 00002370
-rw-r--r--    1 10001    10002     128.0M Dec  4 07:36 00002371
-rw-r--r--    1 10001    10002     128.0M Dec  4 08:15 00002372
-rw-r--r--    1 10001    10002     128.0M Dec  4 08:54 00002373
-rw-r--r--    1 10001    10002     128.0M Dec  4 09:34 00002374
-rw-r--r--    1 10001    10002     107.8M Dec  4 10:07 00002375
-rw-r--r--    1 10001    10002      46.8M Dec  4 10:52 00002376
-rw-r--r--    1 10001    10002      13.7M Dec  4 11:28 00002377
-rw-r--r--    1 10001    10002          0 Dec  4 11:28 00002378
-rw-r--r--    1 10001    10002          0 Dec  4 11:35 00002379
-rw-r--r--    1 10001    10002     105.7M Dec  4 12:28 00002380
-rw-r--r--    1 10001    root           0 Dec  4 12:29 00002381
drwxr-xr-x    2 10001    10002       4.0K Nov 27 17:00 checkpoint.002287
```

**Comment by walterdolce** [comment]
Adding `--storage.tsdb.retention.size=8GB` to the deployment above doesn't make any difference

**Comment by walterdolce** [comment]
Prometheus has been running for more than 2 hours now, yet the ingested samples have not been grouped:
```
prometheus-6997b4b887-c6xvt   1/1     Running   0          140m
```

More WAL files have been added in the meantime..
```
/prometheus/data/wal $ ls -lah
total 10G
drwxr-xr-x    3 10001    10002       4.0K Dec  4 14:39 .
drwxr-xr-x   24 10001    10002       4.0K Dec  4 14:49 ..
-rw-r--r--    1 10001    10002     128.0M Nov 27 14:19 00002288
-rw-r--r--    1 10001    10002     127.9M Nov 27 14:58 00002289
-rw-r--r--    1 10001    10002       4.8M Nov 27 15:00 00002290
-rw-r--r--    1 10001    10002     128.0M Nov 27 15:39 00002291
-rw-r--r--    1 10001    10002     128.0M Nov 27 16:19 00002292
-rw-r--r--    1 10001    10002     128.0M Nov 27 16:58 00002293
-rw-r--r--    1 10001    10002       6.3M Nov 27 17:00 00002294
-rw-r--r--    1 10001    10002      61.1M Nov 27 17:19 00002295
-rw-r--r--    1 10001    10002      23.0M Nov 27 17:24 00002296
-rw-r--r--    1 10001    10002      94.6M Nov 27 17:53 00002297
-rw-r--r--    1 10001    10002     128.0M Nov 27 18:33 00002298
-rw-r--r--    1 10001    10002     127.9M Nov 27 19:13 00002299
-rw-r--r--    1 10001    10002       4.9M Nov 28 13:48 00002300
-rw-r--r--    1 10001    10002     128.0M Nov 28 14:27 00002301
-rw-r--r--    1 10001    10002      10.3M Nov 29 06:24 00002302
-rw-r--r--    1 10001    10002     127.6M Nov 29 07:05 00002303
-rw-r--r--    1 10001    10002     127.9M Nov 29 07:45 00002304
-rw-r--r--    1 10001    10002     128.0M Nov 29 08:26 00002305
-rw-r--r--    1 10001    10002     128.0M Nov 29 09:06 00002306
-rw-r--r--    1 10001    10002     127.9M Nov 29 09:46 00002307
-rw-r--r--    1 10001    10002     127.9M Nov 29 10:26 00002308
-rw-r--r--    1 10001    10002     127.8M Nov 29 11:06 00002309
-rw-r--r--    1 10001    10002     128.0M Nov 29 11:46 00002310
-rw-r--r--    1 10001    10002     127.9M Nov 29 12:26 00002311
-rw-r--r--    1 10001    10002     128.0M Nov 29 13:06 00002312
-rw-r--r--    1 10001    10002     127.7M Nov 29 13:47 00002313
-rw-r--r--    1 10001    10002     128.0M Nov 29 14:27 00002314
-rw-r--r--    1 10001    10002     128.0M Nov 29 15:07 00002315
-rw-r--r--    1 10001    10002     128.0M Nov 29 15:48 00002316
-rw-r--r--    1 10001    10002     127.6M Nov 29 16:28 00002317
-rw-r--r--    1 10001    10002     128.0M Nov 29 17:09 00002318
-rw-r--r--    1 10001    10002     128.0M Nov 29 17:49 00002319
-rw-r--r--    1 10001    10002     128.0M Nov 29 18:29 00002320
-rw-r--r--    1 10001    10002     128.0M Nov 29 19:10 00002321
-rw-r--r--    1 10001    10002     127.6M Nov 29 19:50 00002322
-rw-r--r--    1 10001    10002     128.0M Nov 29 20:31 00002323
-rw-r--r--    1 10001    10002     128.0M Nov 29 21:11 00002324
-rw-r--r--    1 10001    10002      63.6M Nov 29 21:31 00002325
-rw-r--r--    1 10001    10002      68.6M Nov 29 22:08 00002326
-rw-r--r--    1 10001    10002     127.9M Nov 29 22:59 00002327
-rw-r--r--    1 10001    10002     128.0M Nov 29 23:39 00002328
-rw-r--r--    1 10001    10002     128.0M Nov 30 00:20 00002329
-rw-r--r--    1 10001    10002     128.0M Nov 30 01:00 00002330
-rw-r--r--    1 10001    10002     128.0M Nov 30 01:41 00002331
-rw-r--r--    1 10001    10002     127.6M Nov 30 02:21 00002332
-rw-r--r--    1 10001    10002     128.0M Nov 30 03:02 00002333
-rw-r--r--    1 10001    10002     128.0M Nov 30 03:43 00002334
-rw-r--r--    1 10001    10002     127.9M Nov 30 04:24 00002335
-rw-r--r--    1 10001    10002     128.0M Nov 30 05:06 00002336
-rw-r--r--    1 10001    10002     127.7M Nov 30 05:47 00002337
-rw-r--r--    1 10001    10002     128.0M Nov 30 06:28 00002338
-rw-r--r--    1 10001    10002     128.0M Nov 30 07:09 00002339
-rw-r--r--    1 10001    10002     127.9M Dec  3 12:19 00002340
-rw-r--r--    1 10001    10002     128.0M Dec  3 12:58 00002341
-rw-r--r--    1 10001    10002     128.0M Dec  3 13:37 00002342
-rw-r--r--    1 10001    10002     128.0M Dec  3 14:16 00002343
-rw-r--r--    1 10001    10002     127.9M Dec  3 14:55 00002344
-rw-r--r--    1 10001    10002     128.0M Dec  3 15:33 00002345
-rw-r--r--    1 10001    10002     127.9M Dec  3 16:12 00002346
-rw-r--r--    1 10001    10002     128.0M Dec  3 16:50 00002347
-rw-r--r--    1 10001    10002     128.0M Dec  3 17:30 00002348
-rw-r--r--    1 10001    10002      43.2M Dec  3 17:43 00002349
-rw-r--r--    1 10001    10002     127.9M Dec  3 18:40 00002350
-rw-r--r--    1 10001    10002      28.2M Dec  3 18:48 00002351
-rw-r--r--    1 10001    10002          0 Dec  3 18:49 00002352
-rw-r--r--    1 10001    10002      47.5M Dec  3 19:51 00002353
-rw-r--r--    1 10001    10002          0 Dec  3 19:52 00002354
-rw-r--r--    1 10001    10002     128.0M Dec  3 21:17 00002355
-rw-r--r--    1 10001    10002     127.9M Dec  3 21:56 00002356
-rw-r--r--    1 10001    10002     127.9M Dec  3 22:35 00002357
-rw-r--r--    1 10001    10002     128.0M Dec  3 23:13 00002358
-rw-r--r--    1 10001    10002     127.8M Dec  3 23:52 00002359
-rw-r--r--    1 10001    10002     127.9M Dec  4 00:31 00002360
-rw-r--r--    1 10001    10002     128.0M Dec  4 01:09 00002361
-rw-r--r--    1 10001    10002     127.9M Dec  4 01:48 00002362
-rw-r--r--    1 10001    10002     128.0M Dec  4 02:27 00002363
-rw-r--r--    1 10001    10002     127.7M Dec  4 03:06 00002364
-rw-r--r--    1 10001    10002     128.0M Dec  4 03:44 00002365
-rw-r--r--    1 10001    10002     128.0M Dec  4 04:23 00002366
-rw-r--r--    1 10001    10002     127.9M Dec  4 05:02 00002367
-rw-r--r--    1 10001    10002     128.0M Dec  4 05:40 00002368
-rw-r--r--    1 10001    10002     127.8M Dec  4 06:19 00002369
-rw-r--r--    1 10001    10002     128.0M Dec  4 06:58 00002370
-rw-r--r--    1 10001    10002     128.0M Dec  4 07:36 00002371
-rw-r--r--    1 10001    10002     128.0M Dec  4 08:15 00002372
-rw-r--r--    1 10001    10002     128.0M Dec  4 08:54 00002373
-rw-r--r--    1 10001    10002     128.0M Dec  4 09:34 00002374
-rw-r--r--    1 10001    10002     107.8M Dec  4 10:07 00002375
-rw-r--r--    1 10001    10002      46.8M Dec  4 10:52 00002376
-rw-r--r--    1 10001    10002      13.7M Dec  4 11:28 00002377
-rw-r--r--    1 10001    10002          0 Dec  4 11:28 00002378
-rw-r--r--    1 10001    10002          0 Dec  4 11:35 00002379
-rw-r--r--    1 10001    10002     105.7M Dec  4 12:28 00002380
-rw-r--r--    1 10001    root      128.0M Dec  4 13:16 00002381
-rw-r--r--    1 10001    root      128.0M Dec  4 13:58 00002382
-rw-r--r--    1 10001    root      128.0M Dec  4 14:39 00002383
-rw-r--r--    1 10001    root       34.5M Dec  4 14:50 00002384
drwxr-xr-x    2 10001    10002       4.0K Nov 27 17:00 checkpoint.002287
```

**Comment by walterdolce** [comment]
By investigating this further, I found there's +1k `compaction failed` errors reported.

Example:
```
component=tsdb msg="compaction failed" err="persist head block: write compaction: add series: out-of-order series added with label set \"{__name__=\\\"apiserver_audit_event_total\\\",beta_kubernetes_io_arch=\\\"amd64\\\",beta_kubernetes_io_fluentd_ds_ready=\\\"true\\\",beta_kubernetes_io_instance_type=\\\"n1-standard-2\\\",beta_kubernetes_io_os=\\\"linux\\\",cloud_google_com_gke_nodepool=\\\"default-pool\\\",cloud_google_com_gke_os_distribution=\\\"cos\\\",failure_domain_beta_kubernetes_io_region=\\\"europe-west2\\\",failure_domain_beta_kubernetes_io_zone=\\\"europe-west2-a\\\",instance=\\\"gke-bla-bla-kubernetes--default-pool-bla-bla\\\",job=\\\"kubernetes-nodes\\\",kubernetes_io_hostname=\\\"gke-bla-bla-kubernetes--default-pool-bla-bla\\\"}\""
```

**Comment by walterdolce** [comment]
Removing a few `*.tmp` folders did not solve the problem

**Comment by walterdolce** [comment]
After deleting the series above I saw the following being logged:
```
adding interval to ref:  57933
adding interval to ref:  85904
adding interval to ref:  57385
adding interval to ref:  361979
adding interval to ref:  431840
adding interval to ref:  399403
adding interval to ref:  385521
adding interval to ref:  358824
adding interval to ref:  396049
adding interval to ref:  376807
adding interval to ref:  361404
adding interval to ref:  361647
adding interval to ref:  384083
adding interval to ref:  393868
adding interval to ref:  372146
adding interval to ref:  417321
adding interval to ref:  410370
```

After a short while, other `out-of-order series added with label set {{some_set}}` popped up. So I am going to delete those as well as they come up and see what happens...

**Comment by walterdolce** [comment]
I thought I'd add the following just in case they are needed:
```
prometheus_tsdb_head_series: 1232142
prometheus_tsdb_head_series_created_total: 1396409
prometheus_tsdb_head_series_removed_total: 164141
```

I'm still removing series which generate an `out of order ...` error. Same as above. Only different metric.

**Comment by walterdolce** [comment]
By Googling around I found a similar problem is highlighted [here](https://gitlab.com/gitlab-org/omnibus-gitlab/issues/4166). I would like to understand what other mitigation measures can be taken other than "delete and restart".

**Comment by codesome** [comment]
@walterdolce thanks for all the details. `out-of-order series` seems to be the root cause for all this. WAL files are cleaned up only after compaction of head block, which is not happening right now. I am not sure why you are getting out-of-order series for head block, it needs more investigation in the codebase.

(Though this looks like a dupe of #5617, I would like to keep this hanging around)

**Comment by codesome** [comment]
> I would like to understand what other mitigation steps can be taken other than "delete and restart".

I am aware of any other way right now. You would have to (1) Bring down the Prometheus server (2) Delete the WAL (3) Start again - but you would lose a lot of data given how long the data was not compacted from Head. I would suggest you to keep a backup of all the WAL which exists currently before doing it as it's possible to recover data from it and backfill with some effort.

**Comment by walterdolce** [comment]
Hey @codesome, thank you for coming back to me on this.

I left Prometheus running all night since yesterday. It's currently on an instance with much bigger compute capacity on purpose. This should hopefully mitigate any potential problem with the eviction. It's important to note that at the moment I also purposefully removed the specification of `resource.limits` from the deployment spec.

So here's the current status:
```
NAME                          READY   STATUS    RESTARTS   AGE
prometheus-946f576dd-k4cmx    1/1     Running   0          18h
```

Since yesterday, I have manually deleted via the Prometheus API every instance of `out of order series` that popped up in the logs, one by one. 

With regards to:
> I would suggest you to keep a backup of all the WAL which exists currently before doing it as it's possible to recover data from it and backfill with some effort.

I have backed up the entire `wal` directory. I **can** go for the nuclear option and 1) stop, then 2) delete everything and then 3) restart again [by losing all data collected so far].

Two questions I have in relation to recovering the data are:
1) How do you backfill the data? I could not find any documentation for this anywhere
2) Would the files in backed-up `wal` directory clash with the newly created ones? 

Thanks

**Comment by walterdolce** [comment]
Current `wal` folder status. Checkpoint is still the same. Many more `000***` files were added in the meantime. Only an instance of `out of order series` found as I `exec`'d into the container found, which has now been removed via the API.
```
/prometheus/data/wal $ ls -lah
total 14G
drwxr-xr-x    3 10001    10002       4.0K Dec  5 09:09 .
drwxr-xr-x   22 10001    10002       4.0K Dec  5 09:40 ..
-rw-r--r--    1 10001    10002     128.0M Nov 27 14:19 00002288
-rw-r--r--    1 10001    10002     127.9M Nov 27 14:58 00002289
-rw-r--r--    1 10001    10002       4.8M Nov 27 15:00 00002290
-rw-r--r--    1 10001    10002     128.0M Nov 27 15:39 00002291
-rw-r--r--    1 10001    10002     128.0M Nov 27 16:19 00002292
-rw-r--r--    1 10001    10002     128.0M Nov 27 16:58 00002293
-rw-r--r--    1 10001    10002       6.3M Nov 27 17:00 00002294
-rw-r--r--    1 10001    10002      61.1M Nov 27 17:19 00002295
-rw-r--r--    1 10001    10002      23.0M Nov 27 17:24 00002296
-rw-r--r--    1 10001    10002      94.6M Nov 27 17:53 00002297
-rw-r--r--    1 10001    10002     128.0M Nov 27 18:33 00002298
-rw-r--r--    1 10001    10002     127.9M Nov 27 19:13 00002299
-rw-r--r--    1 10001    10002       4.9M Nov 28 13:48 00002300
-rw-r--r--    1 10001    10002     128.0M Nov 28 14:27 00002301
-rw-r--r--    1 10001    10002      10.3M Nov 29 06:24 00002302
-rw-r--r--    1 10001    10002     127.6M Nov 29 07:05 00002303
-rw-r--r--    1 10001    10002     127.9M Nov 29 07:45 00002304
-rw-r--r--    1 10001    10002     128.0M Nov 29 08:26 00002305
-rw-r--r--    1 10001    10002     128.0M Nov 29 09:06 00002306
-rw-r--r--    1 10001    10002     127.9M Nov 29 09:46 00002307
-rw-r--r--    1 10001    10002     127.9M Nov 29 10:26 00002308
-rw-r--r--    1 10001    10002     127.8M Nov 29 11:06 00002309
-rw-r--r--    1 10001    10002     128.0M Nov 29 11:46 00002310
-rw-r--r--    1 10001    10002     127.9M Nov 29 12:26 00002311
-rw-r--r--    1 10001    10002     128.0M Nov 29 13:06 00002312
-rw-r--r--    1 10001    10002     127.7M Nov 29 13:47 00002313
-rw-r--r--    1 10001    10002     128.0M Nov 29 14:27 00002314
-rw-r--r--    1 10001    10002     128.0M Nov 29 15:07 00002315
-rw-r--r--    1 10001    10002     128.0M Nov 29 15:48 00002316
-rw-r--r--    1 10001    10002     127.6M Nov 29 16:28 00002317
-rw-r--r--    1 10001    10002     128.0M Nov 29 17:09 00002318
-rw-r--r--    1 10001    10002     128.0M Nov 29 17:49 00002319
-rw-r--r--    1 10001    10002     128.0M Nov 29 18:29 00002320
-rw-r--r--    1 10001    10002     128.0M Nov 29 19:10 00002321
-rw-r--r--    1 10001    10002     127.6M Nov 29 19:50 00002322
-rw-r--r--    1 10001    10002     128.0M Nov 29 20:31 00002323
-rw-r--r--    1 10001    10002     128.0M Nov 29 21:11 00002324
-rw-r--r--    1 10001    10002      63.6M Nov 29 21:31 00002325
-rw-r--r--    1 10001    10002      68.6M Nov 29 22:08 00002326
-rw-r--r--    1 10001    10002     127.9M Nov 29 22:59 00002327
-rw-r--r--    1 10001    10002     128.0M Nov 29 23:39 00002328
-rw-r--r--    1 10001    10002     128.0M Nov 30 00:20 00002329
-rw-r--r--    1 10001    10002     128.0M Nov 30 01:00 00002330
-rw-r--r--    1 10001    10002     128.0M Nov 30 01:41 00002331
-rw-r--r--    1 10001    10002     127.6M Nov 30 02:21 00002332
-rw-r--r--    1 10001    10002     128.0M Nov 30 03:02 00002333
-rw-r--r--    1 10001    10002     128.0M Nov 30 03:43 00002334
-rw-r--r--    1 10001    10002     127.9M Nov 30 04:24 00002335
-rw-r--r--    1 10001    10002     128.0M Nov 30 05:06 00002336
-rw-r--r--    1 10001    10002     127.7M Nov 30 05:47 00002337
-rw-r--r--    1 10001    10002     128.0M Nov 30 06:28 00002338
-rw-r--r--    1 10001    10002     128.0M Nov 30 07:09 00002339
-rw-r--r--    1 10001    10002     127.9M Dec  3 12:19 00002340
-rw-r--r--    1 10001    10002     128.0M Dec  3 12:58 00002341
-rw-r--r--    1 10001    10002     128.0M Dec  3 13:37 00002342
-rw-r--r--    1 10001    10002     128.0M Dec  3 14:16 00002343
-rw-r--r--    1 10001    10002     127.9M Dec  3 14:55 00002344
-rw-r--r--    1 10001    10002     128.0M Dec  3 15:33 00002345
-rw-r--r--    1 10001    10002     127.9M Dec  3 16:12 00002346
-rw-r--r--    1 10001    10002     128.0M Dec  3 16:50 00002347
-rw-r--r--    1 10001    10002     128.0M Dec  3 17:30 00002348
-rw-r--r--    1 10001    10002      43.2M Dec  3 17:43 00002349
-rw-r--r--    1 10001    10002     127.9M Dec  3 18:40 00002350
-rw-r--r--    1 10001    10002      28.2M Dec  3 18:48 00002351
-rw-r--r--    1 10001    10002          0 Dec  3 18:49 00002352
-rw-r--r--    1 10001    10002      47.5M Dec  3 19:51 00002353
-rw-r--r--    1 10001    10002          0 Dec  3 19:52 00002354
-rw-r--r--    1 10001    10002     128.0M Dec  3 21:17 00002355
-rw-r--r--    1 10001    10002     127.9M Dec  3 21:56 00002356
-rw-r--r--    1 10001    10002     127.9M Dec  3 22:35 00002357
-rw-r--r--    1 10001    10002     128.0M Dec  3 23:13 00002358
-rw-r--r--    1 10001    10002     127.8M Dec  3 23:52 00002359
-rw-r--r--    1 10001    10002     127.9M Dec  4 00:31 00002360
-rw-r--r--    1 10001    10002     128.0M Dec  4 01:09 00002361
-rw-r--r--    1 10001    10002     127.9M Dec  4 01:48 00002362
-rw-r--r--    1 10001    10002     128.0M Dec  4 02:27 00002363
-rw-r--r--    1 10001    10002     127.7M Dec  4 03:06 00002364
-rw-r--r--    1 10001    10002     128.0M Dec  4 03:44 00002365
-rw-r--r--    1 10001    10002     128.0M Dec  4 04:23 00002366
-rw-r--r--    1 10001    10002     127.9M Dec  4 05:02 00002367
-rw-r--r--    1 10001    10002     128.0M Dec  4 05:40 00002368
-rw-r--r--    1 10001    10002     127.8M Dec  4 06:19 00002369
-rw-r--r--    1 10001    10002     128.0M Dec  4 06:58 00002370
-rw-r--r--    1 10001    10002     128.0M Dec  4 07:36 00002371
-rw-r--r--    1 10001    10002     128.0M Dec  4 08:15 00002372
-rw-r--r--    1 10001    10002     128.0M Dec  4 08:54 00002373
-rw-r--r--    1 10001    10002     128.0M Dec  4 09:34 00002374
-rw-r--r--    1 10001    10002     107.8M Dec  4 10:07 00002375
-rw-r--r--    1 10001    10002      46.8M Dec  4 10:52 00002376
-rw-r--r--    1 10001    10002      13.7M Dec  4 11:28 00002377
-rw-r--r--    1 10001    10002          0 Dec  4 11:28 00002378
-rw-r--r--    1 10001    10002          0 Dec  4 11:35 00002379
-rw-r--r--    1 10001    10002     105.7M Dec  4 12:28 00002380
-rw-r--r--    1 10001    10002     128.0M Dec  4 13:16 00002381
-rw-r--r--    1 10001    10002     128.0M Dec  4 13:58 00002382
-rw-r--r--    1 10001    10002     128.0M Dec  4 14:39 00002383
-rw-r--r--    1 10001    10002     105.8M Dec  4 15:13 00002384
-rw-r--r--    1 10001    10002          0 Dec  4 15:13 00002385
-rw-r--r--    1 10001    root      128.0M Dec  4 16:05 00002386
-rw-r--r--    1 10001    root      127.8M Dec  4 16:46 00002387
-rw-r--r--    1 10001    root      128.0M Dec  4 17:27 00002388
-rw-r--r--    1 10001    root      128.0M Dec  4 18:08 00002389
-rw-r--r--    1 10001    root      127.9M Dec  4 18:49 00002390
-rw-r--r--    1 10001    root      127.8M Dec  4 19:30 00002391
-rw-r--r--    1 10001    root      127.8M Dec  4 20:11 00002392
-rw-r--r--    1 10001    root      127.8M Dec  4 20:52 00002393
-rw-r--r--    1 10001    root      127.9M Dec  4 21:33 00002394
-rw-r--r--    1 10001    root      127.8M Dec  4 22:14 00002395
-rw-r--r--    1 10001    root      128.0M Dec  4 22:55 00002396
-rw-r--r--    1 10001    root      127.9M Dec  4 23:36 00002397
-rw-r--r--    1 10001    root      127.9M Dec  5 00:17 00002398
-rw-r--r--    1 10001    root      128.0M Dec  5 00:58 00002399
-rw-r--r--    1 10001    root      128.0M Dec  5 01:39 00002400
-rw-r--r--    1 10001    root      128.0M Dec  5 02:20 00002401
-rw-r--r--    1 10001    root      128.0M Dec  5 03:01 00002402
-rw-r--r--    1 10001    root      128.0M Dec  5 03:42 00002403
-rw-r--r--    1 10001    root      128.0M Dec  5 04:23 00002404
-rw-r--r--    1 10001    root      128.0M Dec  5 05:04 00002405
-rw-r--r--    1 10001    root      128.0M Dec  5 05:45 00002406
-rw-r--r--    1 10001    root      128.0M Dec  5 06:26 00002407
-rw-r--r--    1 10001    root      128.0M Dec  5 07:07 00002408
-rw-r--r--    1 10001    root      128.0M Dec  5 07:48 00002409
-rw-r--r--    1 10001    root      127.7M Dec  5 08:28 00002410
-rw-r--r--    1 10001    root      128.0M Dec  5 09:09 00002411
-rw-r--r--    1 10001    root       97.7M Dec  5 09:41 00002412
drwxr-xr-x    2 10001    10002       4.0K Nov 27 17:00 checkpoint.002287
```

**Comment by kb5zuy** [comment]
I am having a similar problem.  Also running prometheus in a pod in openshift.  I am not seeing any "out of order" messages, but prometheus is not grouping WAL files.  The WAL directory just keeps adding files.  version 2.13.0 did not have this issue; version 2.14+ seem to have the issue.  I need to retest 2.14; 2.15.x has this issue.

      --storage.tsdb.retention.time=24h
      --storage.tsdb.path=/var/prometheus
      --storage.tsdb.min-block-duration=2h
      --storage.tsdb.max-block-duration=2h

(using thanos sidecar for long term storage)

prometheus data directory.  I restarted the pod at 15:37 today.  I see there is a failed .tmp directory.

```
drwxr-xr-x    3 10007400 root          4096 Jan 12 07:00 01DYC8TRCG99TX9AMHXNTVPXY6
drwxr-xr-x    3 10007400 root          4096 Jan 12 09:00 01DYCFPFMFMB69VWEYR4FQ5EKT
drwxr-xr-x    3 10007400 root          4096 Jan 12 11:00 01DYCPJ6WHHN3FBEMQ90TGSY1A
drwxr-xr-x    3 10007400 root          4096 Jan 12 13:00 01DYCXDY4X79JSJ69FHXX2V5KM
drwxr-xr-x    3 10007400 root          4096 Jan 12 15:00 01DYD49ND5SS0W8G695YW4YVKA
drwxr-xr-x    3 10007400 root          4096 Jan 12 17:00 01DYDB5CMHF467YPT9ZQM5PDH6
drwxr-xr-x    3 10007400 root          4096 Jan 12 19:00 01DYDJ13WMQCW8X8MRVGQBDTXS
drwxr-xr-x    3 10007400 root          4096 Jan 12 21:00 01DYDRWV4P5G8VV1TMQNJEEB2K
drwxr-xr-x    3 10007400 root          4096 Jan 12 23:00 01DYDZRJCE97YNCVWQDDS5FT4R
drwxr-xr-x    3 10007400 root          4096 Jan 13 01:00 01DYE6M9ME0NBD831JGPZ510GY
drwxr-xr-x    3 10007400 root          4096 Jan 13 03:00 01DYEDG0WE1SNEP0Z27WRJ0S33
drwxr-xr-x    3 10007400 root          4096 Jan 13 05:00 01DYEMBR4EG9SBWCDX75316JF7
drwxr-xr-x    3 10007400 root          4096 Jan 13 07:00 01DYEV7FCKMYS55VBP5KWVQ8HM
drwxr-xr-x    3 10007400 root          4096 Jan 14 15:38 01DYJBA6GJEF83SB5H5VQ0HXJ2.tmp
-rw-r--r--    1 10007400 root             0 May 13  2019 lock
-rw-r--r--    1 10007400 root         20001 Jan 14 15:37 queries.active
drwxr-xr-x    4 10007400 root          4096 Jan 14 15:56 snapshots
drwxr-xr-x    3 10007400 root          4096 May 13  2019 thanos
-rw-r--r--    1 10007400 root           452 Jan 14 16:28 thanos.shipper.json
drwxr-xr-x    3 10007400 root          4096 Jan 14 15:37 wal

```

wal directory:
```
total 4597483
-rw-r--r--    1 10007400 root     112164864 Jan 13 03:00 00000111
-rw-r--r--    1 10007400 root     134217728 Jan 13 04:05 00000112
-rw-r--r--    1 10007400 root     112197632 Jan 13 05:00 00000113
-rw-r--r--    1 10007400 root     134217728 Jan 13 06:05 00000114
-rw-r--r--    1 10007400 root     112197632 Jan 13 07:00 00000115
-rw-r--r--    1 10007400 root     102727680 Jan 13 07:49 00000116
-rw-r--r--    1 10007400 root     134217728 Jan 13 08:55 00000117
-rw-r--r--    1 10007400 root     134217728 Jan 13 10:00 00000118
-rw-r--r--    1 10007400 root     134184960 Jan 13 11:04 00000119
-rw-r--r--    1 10007400 root     134217728 Jan 13 12:09 00000120
-rw-r--r--    1 10007400 root     134184960 Jan 13 13:14 00000121
-rw-r--r--    1 10007400 root     134217728 Jan 13 14:19 00000122
-rw-r--r--    1 10007400 root     134217728 Jan 13 15:24 00000123
-rw-r--r--    1 10007400 root     134217728 Jan 13 16:29 00000124
-rw-r--r--    1 10007400 root     134217728 Jan 13 17:34 00000125
-rw-r--r--    1 10007400 root     134217728 Jan 13 18:39 00000126
-rw-r--r--    1 10007400 root     134217728 Jan 13 19:44 00000127
-rw-r--r--    1 10007400 root     134217728 Jan 13 20:49 00000128
-rw-r--r--    1 10007400 root     134217728 Jan 13 21:55 00000129
-rw-r--r--    1 10007400 root     134217728 Jan 13 23:00 00000130
-rw-r--r--    1 10007400 root     134217728 Jan 14 00:06 00000131
-rw-r--r--    1 10007400 root     134217728 Jan 14 01:11 00000132
-rw-r--r--    1 10007400 root     134217728 Jan 14 02:16 00000133
-rw-r--r--    1 10007400 root     134217728 Jan 14 03:22 00000134
-rw-r--r--    1 10007400 root     134184960 Jan 14 04:27 00000135
-rw-r--r--    1 10007400 root     134217728 Jan 14 05:33 00000136
-rw-r--r--    1 10007400 root     134184960 Jan 14 06:38 00000137
-rw-r--r--    1 10007400 root     134217728 Jan 14 07:43 00000138
-rw-r--r--    1 10007400 root     134217728 Jan 14 08:48 00000139
-rw-r--r--    1 10007400 root     134217728 Jan 14 09:52 00000140
-rw-r--r--    1 10007400 root     134217728 Jan 14 10:57 00000141
-rw-r--r--    1 10007400 root     134217728 Jan 14 12:02 00000142
-rw-r--r--    1 10007400 root     134184960 Jan 14 13:06 00000143
-rw-r--r--    1 10007400 root     134217728 Jan 14 14:11 00000144
-rw-r--r--    1 10007400 root     134217728 Jan 14 15:16 00000145
-rw-r--r--    1 10007400 root      40828928 Jan 14 15:36 00000146
-rw-r--r--    1 10007400 root      67115119 Jan 14 16:11 00000147
drwxr-xr-x    2 10007400 root          4096 Jan 13 07:00 checkpoint.000110

```

**Comment by gerardo8a** [comment]
I don't want to steal the github issue, but I've been looking for some solution and I think we have a similar issue. If needed I can open a new issue. [@codesome]

I have the something similar, the heap chunks keep growing until the pod gets OOMKilled. I'm using 224G of memory and have the latest prometheus version v2.15.2

This the message I get when it gets OOMKilled
```level=info ts=2020-02-06T22:00:21.181Z caller=head.go:668 component=tsdb msg="head GC completed" duration=10.772841355s
level=error ts=2020-02-06T22:03:10.131Z caller=db.go:609 component=tsdb msg="compaction failed" err="head truncate failed (in compact): create checkpoint: flush records: write /var/prometheus/data/wal/checkpoint.000247.tmp/00000025: cannot allocate memory"``` 

This is hot the WAL directory looks like
```-rw-r--r-- 1 nfsnobody nfsnobody 133758976 Feb  6 21:33 00000296
-rw-r--r-- 1 nfsnobody nfsnobody 129499136 Feb  6 21:34 00000297
-rw-r--r-- 1 nfsnobody nfsnobody 126681088 Feb  6 21:34 00000298
-rw-r--r-- 1 nfsnobody nfsnobody  92521340 Feb  6 21:34 00000299
-rw-r--r-- 1 nfsnobody nfsnobody         0 Feb  6 21:34 00000300
drwxr-xr-x 2 nfsnobody nfsnobody     12288 Feb  6 20:11 checkpoint.000218
drwxr-xr-x 2 nfsnobody nfsnobody      8192 Feb  6 20:20 checkpoint.000231.tmp
drwxr-xr-x 2 nfsnobody nfsnobody      4096 Feb  6 20:34 checkpoint.000234.tmp
drwxr-xr-x 2 nfsnobody nfsnobody      8192 Feb  6 20:52 checkpoint.000236.tmp
drwxr-xr-x 2 nfsnobody nfsnobody      4096 Feb  6 21:06 checkpoint.000240.tmp
drwxr-xr-x 2 nfsnobody nfsnobody      4096 Feb  6 21:20 checkpoint.000242.tmp
drwxr-xr-x 2 nfsnobody nfsnobody      4096 Feb  6 21:34 checkpoint.000244.tmp```

My prometheus configuration
``` --config.file=/etc/prometheus/config/prometheus-federation.yaml
    - --storage.tsdb.path=/var/prometheus/data
    - --storage.tsdb.retention.size=4TB
    - --storage.tsdb.max-block-duration=60m
    - --storage.tsdb.min-block-duration=30m
    - --log.level=debug
    - --web.route-prefix=/
    - --web.enable-lifecycle
    - --web.enable-admin-api
    - --web.max-connections=10000
    - --web.external-url=https://mydomain.com:4443/
    - --query.max-concurrency=512
    - --query.timeout=5m```

Heap Chunk never go down, compared to other instances I have and they go up and after some time it goes down, I have 8 more instances and they all behave the same way. This instances is a federated one.
![Screen Shot 2020-02-06 at 12 08 32 PM](https://user-images.githubusercontent.com/217191/73982065-6b6adf00-48e8-11ea-9f0a-825c297c2c59.png)

After this happens it keep crashing.
![Screen Shot 2020-02-06 at 1 59 40 PM](https://user-images.githubusercontent.com/217191/73982393-fd72e780-48e8-11ea-9d07-ac073df9671f.png)

We are using a filer (PV, PVC) for persisten storage.

Here is the pprof output file before it crashed.
[heap02-06-2020-12-10.out.gz](https://github.com/prometheus/prometheus/files/4167972/heap02-06-2020-12-10.out.gz)

**Comment by codesome** [comment]
@gerardo8a given you have `storage.tsdb.min-block-duration=30m`, I don't see compaction happening in your dashboard, which can be explained by the error message that was shown about not enough memory. No compaction is the reason it is just growing. Checkpoints are created after compaction and as checkpointing is failing, the in-memory block (Head) truncation is also failing. Can you try throwing in some more memory and see if it works fine?


PS:  your issue is different than the one described in this issue.

**Comment by Quicksy** [comment]
Same issue here. Reproducible in both 2.15.2 and 2.16.0. Prometheus ran in docker-compose. The WAL directory grows and eats all the space on the host machine. When I restart the container, it frees up to 20+GBs, lose data for last ~week  and the process starts once again.

**Comment by stefanandres** [comment]
We have the same with prometheus (prometheus-1.15 with stable/prometheus-operator-8.12.3)

We are using replicas: 2 and randomly one of the prometheus instance will endlessly grow the wal files.
When we restart the prometheus pod, the wal files are deleted and we loose all data in them.


This happens right now for example:
```
/prometheus/wal $ ls -la1
total 19343648
drwxrwsr-x    4 1000     2000         20480 Apr  8 13:48 .
drwxrwsr-x   33 root     2000          4096 Apr  6 11:00 ..
-rw-r--r--    1 1000     2000        851968 Apr  6 08:10 00002545
-rw-r--r--    1 1000     2000        589824 Apr  6 08:10 00002546
-rw-r--r--    1 1000     2000     133955584 Apr  6 08:33 00002547
-rw-r--r--    1 1000     2000     134152192 Apr  6 08:55 00002548
-rw-r--r--    1 1000     2000      29032448 Apr  6 09:00 00002549
-rw-r--r--    1 1000     2000     134217728 Apr  6 09:22 00002550
-rw-r--r--    1 1000     2000     134119424 Apr  6 09:45 00002551
-rw-r--r--    1 1000     2000     134184960 Apr  6 10:07 00002552
-rw-r--r--    1 1000     2000     134152192 Apr  6 10:29 00002553
-rw-r--r--    1 1000     2000     134184960 Apr  6 10:52 00002554
-rw-r--r--    1 1000     2000     134184960 Apr  6 11:14 00002555
-rw-r--r--    1 1000     2000     134217728 Apr  6 11:36 00002556
-rw-r--r--    1 1000     2000     134217728 Apr  6 11:59 00002557
-rw-r--r--    1 1000     2000     134217728 Apr  6 12:21 00002558
-rw-r--r--    1 1000     2000     134217728 Apr  6 12:44 00002559
-rw-r--r--    1 1000     2000     134184960 Apr  6 13:06 00002560
-rw-r--r--    1 1000     2000     134119424 Apr  6 13:29 00002561
-rw-r--r--    1 1000     2000     134152192 Apr  6 13:51 00002562
-rw-r--r--    1 1000     2000     134217728 Apr  6 14:14 00002563
-rw-r--r--    1 1000     2000     134217728 Apr  6 14:36 00002564
-rw-r--r--    1 1000     2000     134217728 Apr  6 14:59 00002565
-rw-r--r--    1 1000     2000     134217728 Apr  6 15:21 00002566
-rw-r--r--    1 1000     2000     133857280 Apr  6 15:44 00002567
-rw-r--r--    1 1000     2000     134184960 Apr  6 16:06 00002568
-rw-r--r--    1 1000     2000     134217728 Apr  6 16:28 00002569
-rw-r--r--    1 1000     2000     134217728 Apr  6 16:51 00002570
-rw-r--r--    1 1000     2000     134217728 Apr  6 17:13 00002571
-rw-r--r--    1 1000     2000     134184960 Apr  6 17:36 00002572
-rw-r--r--    1 1000     2000     134184960 Apr  6 17:58 00002573
-rw-r--r--    1 1000     2000     134217728 Apr  6 18:21 00002574
-rw-r--r--    1 1000     2000     134217728 Apr  6 18:43 00002575
-rw-r--r--    1 1000     2000     134184960 Apr  6 19:06 00002576
-rw-r--r--    1 1000     2000     134217728 Apr  6 19:28 00002577
-rw-r--r--    1 1000     2000     134119424 Apr  6 19:51 00002578
-rw-r--r--    1 1000     2000     134217728 Apr  6 20:13 00002579
-rw-r--r--    1 1000     2000     134184960 Apr  6 20:35 00002580
-rw-r--r--    1 1000     2000     134152192 Apr  6 20:58 00002581
-rw-r--r--    1 1000     2000     134217728 Apr  6 21:20 00002582
-rw-r--r--    1 1000     2000     134217728 Apr  6 21:43 00002583
-rw-r--r--    1 1000     2000     134217728 Apr  6 22:05 00002584
-rw-r--r--    1 1000     2000     133955584 Apr  6 22:28 00002585
-rw-r--r--    1 1000     2000     134217728 Apr  6 22:50 00002586
-rw-r--r--    1 1000     2000     134119424 Apr  6 23:12 00002587
-rw-r--r--    1 1000     2000     134152192 Apr  6 23:35 00002588
-rw-r--r--    1 1000     2000     134021120 Apr  6 23:57 00002589
-rw-r--r--    1 1000     2000     133726208 Apr  7 00:20 00002590
-rw-r--r--    1 1000     2000     134217728 Apr  7 00:42 00002591
-rw-r--r--    1 1000     2000     134217728 Apr  7 01:04 00002592
-rw-r--r--    1 1000     2000     134119424 Apr  7 01:27 00002593
-rw-r--r--    1 1000     2000     134053888 Apr  7 01:49 00002594
-rw-r--r--    1 1000     2000     134184960 Apr  7 02:12 00002595
-rw-r--r--    1 1000     2000     133890048 Apr  7 02:34 00002596
-rw-r--r--    1 1000     2000     134184960 Apr  7 02:56 00002597
-rw-r--r--    1 1000     2000     134217728 Apr  7 03:19 00002598
-rw-r--r--    1 1000     2000     134184960 Apr  7 03:41 00002599
-rw-r--r--    1 1000     2000     134217728 Apr  7 04:04 00002600
-rw-r--r--    1 1000     2000     134217728 Apr  7 04:26 00002601
-rw-r--r--    1 1000     2000     133890048 Apr  7 04:49 00002602
-rw-r--r--    1 1000     2000     134217728 Apr  7 05:11 00002603
-rw-r--r--    1 1000     2000     134184960 Apr  7 05:33 00002604
-rw-r--r--    1 1000     2000     134184960 Apr  7 05:56 00002605
-rw-r--r--    1 1000     2000     134217728 Apr  7 06:18 00002606
-rw-r--r--    1 1000     2000     134217728 Apr  7 06:40 00002607
-rw-r--r--    1 1000     2000     134217728 Apr  7 07:03 00002608
-rw-r--r--    1 1000     2000     133791744 Apr  7 07:25 00002609
-rw-r--r--    1 1000     2000     134184960 Apr  7 07:47 00002610
-rw-r--r--    1 1000     2000     134217728 Apr  7 08:09 00002611
-rw-r--r--    1 1000     2000     133988352 Apr  7 08:32 00002612
-rw-r--r--    1 1000     2000     134184960 Apr  7 08:54 00002613
-rw-r--r--    1 1000     2000     134217728 Apr  7 09:15 00002614
-rw-r--r--    1 1000     2000     134217728 Apr  7 09:38 00002615
-rw-r--r--    1 1000     2000     134217728 Apr  7 10:00 00002616
-rw-r--r--    1 1000     2000     133758976 Apr  7 10:22 00002617
-rw-r--r--    1 1000     2000     134217728 Apr  7 10:44 00002618
-rw-r--r--    1 1000     2000     134217728 Apr  7 11:06 00002619
-rw-r--r--    1 1000     2000     134152192 Apr  7 11:29 00002620
-rw-r--r--    1 1000     2000     134217728 Apr  7 11:51 00002621
-rw-r--r--    1 1000     2000     133791744 Apr  7 12:13 00002622
-rw-r--r--    1 1000     2000     134217728 Apr  7 12:35 00002623
-rw-r--r--    1 1000     2000     134217728 Apr  7 12:57 00002624
-rw-r--r--    1 1000     2000     134217728 Apr  7 13:19 00002625
-rw-r--r--    1 1000     2000     133857280 Apr  7 13:42 00002626
-rw-r--r--    1 1000     2000     134217728 Apr  7 14:03 00002627
-rw-r--r--    1 1000     2000     133660672 Apr  7 14:24 00002628
-rw-r--r--    1 1000     2000     134217728 Apr  7 14:45 00002629
-rw-r--r--    1 1000     2000     134217728 Apr  7 15:07 00002630
-rw-r--r--    1 1000     2000     134152192 Apr  7 15:29 00002631
-rw-r--r--    1 1000     2000     134217728 Apr  7 15:51 00002632
-rw-r--r--    1 1000     2000     134217728 Apr  7 16:13 00002633
-rw-r--r--    1 1000     2000     134184960 Apr  7 16:35 00002634
-rw-r--r--    1 1000     2000     134217728 Apr  7 16:57 00002635
-rw-r--r--    1 1000     2000     134217728 Apr  7 17:19 00002636
-rw-r--r--    1 1000     2000     133955584 Apr  7 17:42 00002637
-rw-r--r--    1 1000     2000     134217728 Apr  7 18:04 00002638
-rw-r--r--    1 1000     2000     134217728 Apr  7 18:26 00002639
-rw-r--r--    1 1000     2000     133988352 Apr  7 18:48 00002640
-rw-r--r--    1 1000     2000     134184960 Apr  7 19:10 00002641
-rw-r--r--    1 1000     2000     134119424 Apr  7 19:32 00002642
-rw-r--r--    1 1000     2000     133922816 Apr  7 19:55 00002643
-rw-r--r--    1 1000     2000     134217728 Apr  7 20:17 00002644
-rw-r--r--    1 1000     2000     134217728 Apr  7 20:39 00002645
-rw-r--r--    1 1000     2000     134184960 Apr  7 21:01 00002646
-rw-r--r--    1 1000     2000     134217728 Apr  7 21:23 00002647
-rw-r--r--    1 1000     2000     134184960 Apr  7 21:45 00002648
-rw-r--r--    1 1000     2000     134217728 Apr  7 22:07 00002649
-rw-r--r--    1 1000     2000     134184960 Apr  7 22:30 00002650
-rw-r--r--    1 1000     2000     134217728 Apr  7 22:52 00002651
-rw-r--r--    1 1000     2000     133791744 Apr  7 23:14 00002652
-rw-r--r--    1 1000     2000     133922816 Apr  7 23:36 00002653
-rw-r--r--    1 1000     2000     134217728 Apr  7 23:58 00002654
-rw-r--r--    1 1000     2000     134217728 Apr  8 00:20 00002655
-rw-r--r--    1 1000     2000     134184960 Apr  8 00:42 00002656
-rw-r--r--    1 1000     2000     134217728 Apr  8 01:04 00002657
-rw-r--r--    1 1000     2000     134184960 Apr  8 01:26 00002658
-rw-r--r--    1 1000     2000     134217728 Apr  8 01:48 00002659
-rw-r--r--    1 1000     2000     133824512 Apr  8 02:11 00002660
-rw-r--r--    1 1000     2000     134053888 Apr  8 02:33 00002661
-rw-r--r--    1 1000     2000     134184960 Apr  8 02:55 00002662
-rw-r--r--    1 1000     2000     134184960 Apr  8 03:17 00002663
-rw-r--r--    1 1000     2000     134119424 Apr  8 03:39 00002664
-rw-r--r--    1 1000     2000     134217728 Apr  8 04:01 00002665
-rw-r--r--    1 1000     2000     134217728 Apr  8 04:23 00002666
-rw-r--r--    1 1000     2000     133988352 Apr  8 04:45 00002667
-rw-r--r--    1 1000     2000     134217728 Apr  8 05:07 00002668
-rw-r--r--    1 1000     2000     134217728 Apr  8 05:29 00002669
-rw-r--r--    1 1000     2000     133922816 Apr  8 05:51 00002670
-rw-r--r--    1 1000     2000     134184960 Apr  8 06:13 00002671
-rw-r--r--    1 1000     2000     134217728 Apr  8 06:35 00002672
-rw-r--r--    1 1000     2000     134217728 Apr  8 06:57 00002673
-rw-r--r--    1 1000     2000     133988352 Apr  8 07:20 00002674
-rw-r--r--    1 1000     2000     134086656 Apr  8 07:42 00002675
-rw-r--r--    1 1000     2000     133726208 Apr  8 08:04 00002676
-rw-r--r--    1 1000     2000     134217728 Apr  8 08:26 00002677
-rw-r--r--    1 1000     2000     134217728 Apr  8 08:48 00002678
-rw-r--r--    1 1000     2000     134217728 Apr  8 09:10 00002679
-rw-r--r--    1 1000     2000     134217728 Apr  8 09:32 00002680
-rw-r--r--    1 1000     2000     134184960 Apr  8 09:54 00002681
-rw-r--r--    1 1000     2000     134217728 Apr  8 10:15 00002682
-rw-r--r--    1 1000     2000     133955584 Apr  8 10:37 00002683
-rw-r--r--    1 1000     2000     134184960 Apr  8 10:59 00002684
-rw-r--r--    1 1000     2000     134217728 Apr  8 11:21 00002685
-rw-r--r--    1 1000     2000     133890048 Apr  8 11:43 00002686
-rw-r--r--    1 1000     2000     134217728 Apr  8 12:04 00002687
-rw-r--r--    1 1000     2000     134217728 Apr  8 12:26 00002688
-rw-r--r--    1 1000     2000     133660672 Apr  8 12:46 00002689
-rw-r--r--    1 1000     2000     134184960 Apr  8 12:59 00002690
-rw-r--r--    1 1000     2000     134184960 Apr  8 13:11 00002691
-rw-r--r--    1 1000     2000     134119424 Apr  8 13:23 00002692
-rw-r--r--    1 1000     2000     134217728 Apr  8 13:36 00002693
-rw-r--r--    1 1000     2000     134217728 Apr  8 13:48 00002694
-rw-r--r--    1 1000     2000      59023571 Apr  8 13:53 00002695
drwxr-sr-x    2 1000     2000          4096 Apr  6 09:00 checkpoint.002544
drwxrwsr-x    2 1000     2000          4096 Mar 19 09:00 checkpoint.012999
/prometheus/wal $
```

**Comment by brian-brazil** [comment]
@stefanandres That looks like an unrelated issue, likely due to having Prometheus servers sharing a directory (which isn't possible). 

@gerardo8a  That looks like an unrelated issue, likely due to using NFS which is not supported.

Unless your issue is identical in every way and you can assist in debugging, please don't add confusion to existing bugs.

**Comment by stefanandres** [comment]
> @stefanandres That looks like an unrelated issue, likely due to having Prometheus servers sharing a directory (which isn't possible).

@brian-brazil My prometheus instances do not share directories. When using the prometheus-operator with one prometheus object and replicas: 2 it will create a StatefulSet (meaning, each instance has it's own volume).  
Each prometheus is completely on it's own and for me this still seems like the same issue as the requester.

In my case, one prometheus instance was OOM'ed and then the described WAL file behavior started.

I'm not sure how this does not match the describing ticket, can you elaborate?

**Comment by brian-brazil** [comment]
@stefanandres You have a checkpoint with an index far ahead of everything else, thus it is a different issue.

**Comment by stefanandres** [comment]
@brian-brazil Thanks for the hint. I guess I see what you mean. The log messages I was showing were from *both* prometheus instances (I've was missing to filter for the *one* instance in the Loki logging server). That could be misread as a shared instance, but I assure you - it is not.

I've updated https://gist.github.com/stefanandres/7280128c8ac3ba0e6f0ce836c5e00cbf to only include log messages from the prometheus-1 instance that showed the behavior. Do you still think this is another issue?

**Comment by brian-brazil** [comment]
Yes, it's still a different set of symptoms. Please stop adding confusion to this one.

**Comment by stefanandres** [comment]
Even by adding a chance of more "annoyance" with my new (and last comment in this issue) comment, I will take the time to still thank you.

Thank you for taking your time by looking into our set of prometheus issues which looked the same at first.

Even though your communication style on my pior comments feels kind of harsh and unfair (me just trying to understand the issue) to me, I greatly appreciate your work.

**Comment by brian-brazil** [comment]
Apologies if I came across as harsh, having bugs descend into 5+ distinct support requests is unworkable from our end.

@walterdolce Are you still seeing this issue with the latest version? If so can you please share logs of the failing compaction?

**Comment by walterdolce** [comment]
@brian-brazil we are no longer using the rolling upgrade strategy as highlighted in the `Deployment` at the beginning of this issue. It was too unreliable for us and we did not want to risk having to play this game every so often.

**Comment by brian-brazil** [comment]
If we can't reproduce then I'm going to have to guess it was something specific to your system, as this is an error that shouldn't happen - even with data corruption on disk. The relevant code has also been reworked since then, so I'll close this for now.

**Comment by roidelapluie** [comment]
If you have 2 you Prometheus running against the same storage that will create issues.

**Comment by ekeih** [comment]
@roidelapluie I think nobody in this issue had two Prometheus instances running with the same storage. That was a misinterpretation of the output in https://github.com/prometheus/prometheus/issues/6408#issuecomment-610981001.
@brian-brazil You say it looks still like a different "set of symptoms". Could you elaborate on that? What looks different? With this information we maybe would be able to create a new issue.
I understand that you have a high volume of issues in this project and can't spend to much time on each, but short "This isn't the same issue" without even a short explanation why, doesn't really help. Because for me, without a deep understanding of the storage format, it looks like the same issue.

**Comment by Quicksy** [comment]
And still I have the same issue as TS with latest version 0.17.1

**Comment by brian-brazil** [comment]
We have now actually reached 5 distinct support requests in one (closed) issue, so if you believe you have found a new bug in Prometheus please open a new issue.

**Comment by hyserver** [comment]
@brian-brazil Can you take a look at it for me? I had the same problem, promethues version：2.18, Version 2.19 is currently unstable，cannot update.
wal]# ll
total 69248328
-rw-r--r-- 1 prometheus prometheus 129269760 Jun 15 12:45 00000134
-rw-r--r-- 1 prometheus prometheus 134021120 Jun 15 12:45 00000135
-rw-r--r-- 1 prometheus prometheus 133562368 Jun 15 12:45 00000136
-rw-r--r-- 1 prometheus prometheus 132874240 Jun 15 12:46 00000137
-rw-r--r-- 1 prometheus prometheus 132743168 Jun 15 12:46 00000138
-rw-r--r-- 1 prometheus prometheus 132775936 Jun 15 12:46 00000139
-rw-r--r-- 1 prometheus prometheus 133398528 Jun 15 12:46 00000140
-rw-r--r-- 1 prometheus prometheus 133103616 Jun 15 12:47 00000141
-rw-r--r-- 1 prometheus prometheus 132972544 Jun 15 12:47 00000142
-rw-r--r-- 1 prometheus prometheus 133464064 Jun 15 12:47 00000143
-rw-r--r-- 1 prometheus prometheus 132841472 Jun 15 12:47 00000144
-rw-r--r-- 1 prometheus prometheus 131694592 Jun 15 12:48 00000145
-rw-r--r-- 1 prometheus prometheus 134086656 Jun 15 12:48 00000146
-rw-r--r-- 1 prometheus prometheus 133300224 Jun 15 12:48 00000147
-rw-r--r-- 1 prometheus prometheus 133103616 Jun 15 12:48 00000148
-rw-r--r-- 1 prometheus prometheus 132808704 Jun 15 12:49 00000149
-rw-r--r-- 1 prometheus prometheus 132874240 Jun 15 12:49 00000150
-rw-r--r-- 1 prometheus prometheus 133038080 Jun 15 12:49 00000151
-rw-r--r-- 1 prometheus prometheus 132579328 Jun 15 12:49 00000152
-rw-r--r-- 1 prometheus prometheus 133234688 Jun 15 12:50 00000153
-rw-r--r-- 1 prometheus prometheus 132677632 Jun 15 12:50 00000154

**Comment by johns007n** [comment]
@hyserver Did above issue resolved?

If yes then can you share here how to fix
